tests/test_tools.py

```diff
@@ -728,3 +730,51 @@ def test_group_member_ids(input, expected):
     actual = group_member_ids(input, "bg")
     assert actual == expected
 
+p1 = {
+    "_id": "scopatz",
+    "aka": [
+        "Scopatz",
+        "Scopatz, A",
+        "Scopatz, A.",
+        "Scopatz, A M",
+        "Anthony Michael Scopatz",
+    ],
+    "name": "Anthony Scopatz",
+}
+p2 = {
+    "_id": "abc",
+    "aka": [
+        "A. BC",
+        "BC, A",
+        "Anthony BC",
+    ],
+    "name": "Anthony Bill Chris",
+}
+
+@pytest.mark.parametrize(
+    "input, expected",
+    [
+        (([p1, p2], ["aka", "name", "_id"],
+                           "Anth", False),[p1,p2]),
+        (([p1, p2], ["aka", "name", "_id"],
+                           "scopatz, a", True),[]),
+        (([p1, p2], ["aka", "name", "_id"],
+                           "scopatz, a", False),[p1]),
+        (([p1, p2], ["aka", "name", "_id"],
+                           "ill", False),[p2]),          
+    ],
+)
+def test_fragment_retrieval(input, expected):
+    assert(fragment_retrieval(input[0],input[1],input[2],input[3]) == expected)
+
+@pytest.mark.parametrize(
+    "input, expected",
+    [
+        ((None, None), "present"),
+        ((None, 2002), "2002"),
+        ((5,2002), "May 2002"),        
+    ],
+)
+def test_month_and_year(input,expected):
+    assert(month_and_year(input[0],input[1]) == expected)
+    
```

Comment:
there is some weird line-end issue here that I don't understand.  It could cause problems later so if you can figure it out and fix it would be good.

Reply:
I thought there was supposed to be a blank line at the end of the file, but I could be misinformed about that.